### PR TITLE
Chapter 14: Following users

### PR DESCRIPTION
**Pull request by baaohuii**

### CODE REVIEW

**Comment by baaohuii** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -100,6 +100,9 @@ aside {
   margin-top: 15px;
 }
 
+
+
+
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]: remove

File: app/assets/stylesheets/relationships.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Relationships controller here.
```
Comment (chungpv-1008) [inline app/assets/stylesheets/relationships.scss]: remove this file

File: app/controllers/relationships_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    relationship = Relationship.find_by id: params[:id]
+    if relationship.present?
+      @user = relationship.followed
+      current_user.unfollow @user
+    else
+      @error = t "follow.action_fail"
+    end
+    respond_to :js
+      format.js
+    end
```
Comment (chungpv-1008) [inline app/controllers/relationships_controller.rb]:
```suggestion
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 class UsersController < ApplicationController
   before_action :load_user, only: %i(edit update destroy)
-  before_action :logged_in_user, only: %i(show new create)
+  before_action :logged_in_user, only: %i(index edit update destroy following followers)
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
  before_action :logged_in_user, except: %i(new create)
```

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -45,6 +45,20 @@ def destroy
     flash[:success] = "User deleted"
     redirect_to users_url
   end
+  
+  def following
+    @title = "Following"
+    @user = User.find params[:id]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    @user = User.find_by id: params[:id]
```

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -45,6 +45,20 @@ def destroy
...
+    @users = @user.following.paginate page: params[:page]
+    render :show_follow
+  end
+
+  def followers
+    @title = "Followers"
+    @user = User.find params[:id]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    @user = User.find_by id: params[:id]
```

File: app/models/relationship.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+class Relationship < ApplicationRecord
+  belongs_to :follower, class_name: "User"
+  belongs_to :followed, class_name: "User"
```
Comment (chungpv-1008) [inline app/models/relationship.rb]:
```suggestion
  belongs_to :follower, class_name: User.name
  belongs_to :followed, class_name: User.name
```

File: app/models/user.rb [hunk 8]
```diff
@@ -9,20 +9,22 @@ class User < ApplicationRecord
...
+            allow_nil: true
 
   before_save :downcase_email
   before_create :create_activation_digest
   has_secure_password
   has_many :microposts, dependent: :destroy
+  has_many :active_relationships, class_name: "Relationship",
+            foreign_key: "follower_id", dependent: :destroy
+  has_many :passive_relationships, class_name: "Relationship",
+            foreign_key: "followed_id", dependent: :destroy
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
  has_many :active_relationships, class_name: Relationship.name,
            foreign_key: :follower_id, dependent: :destroy
  has_many :passive_relationships, class_name: Relationship.name,
            foreign_key: :followed_id, dependent: :destroy
```

File: app/models/user.rb [hunk 9]
```diff
@@ -79,14 +81,27 @@ def send_activation_email
   end
   
   def feed
-    Micropost.where user_id: id
+    following_ids = "SELECT followed_id FROM relationships WHERE follower_id = :user_id"
+    Micropost.where "user_id IN (#{following_ids}) OR user_id = :user_id", user_id: id
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
    Micropost.where user_id: followed_ids << id
```

File: app/views/relationships/create.js.erb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+$("#follow_form").html("<%= escape_javascript(render("users/unfollow")) %>");
```
Comment (chungpv-1008) [inline app/views/relationships/create.js.erb]: using ' instead of '

File: app/views/relationships/create.js.erb [hunk 11]
```diff
@@ -0,0 +1,2 @@
...
+$("#followers").html("<%= @user.followers.count %>");
```
Comment (chungpv-1008) [inline app/views/relationships/create.js.erb]:
```suggestion
$("#followers").html("<%= @user.followers.size %>");
```

File: app/views/relationships/destroy.js.erb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+$("#follow_form").html("<%= escape_javascript(render("users/follow")) %>");
+$("#followers").html("<%= @user.followers.count %>");
```
Comment (chungpv-1008) [inline app/views/relationships/destroy.js.erb]: same

File: app/views/shared/_user_info.html.erb [hunk 13]
```diff
@@ -1,4 +1,4 @@
-<%= link_to gravatar_for(current_user, size: 50), current_user %>
+<%= link_to gravatar_for (current_user, size= Settings.gravatar_size), current_user %>
```
Comment (chungpv-1008) [inline app/views/shared/_user_info.html.erb]:
```suggestion
<%= link_to gravatar_for(current_user, size= Settings.gravatar_size), current_user %>
```

File: config/settings.yml [hunk 14]
```diff
@@ -21,8 +21,8 @@ validations:
     max_size_Mb: 5
     size: 1024
     resize_limit: 500
-    formats: image/jpeg,image/gif,image/png
-    format_type: %w[image/jpeg image/gif image/png]
+    formats: 'image/jpeg,image/gif,image/png'
```
Comment (chungpv-1008) [inline config/settings.yml]: using " instead of

**Comment by baaohuii** [comment]
ready

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -1,7 +1,7 @@
 class UsersController < ApplicationController
   before_action :load_user, only: %i(edit update destroy)
-  before_action :logged_in_user, only: %i(show new create)
-  before_action :correct_user, only: %i(edit update)
+  before_action :logged_in_user, only: %i(index edit update destroy following followers)
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]: Not fix yet

File: app/views/relationships/create.js.erb [hunk 16]
```diff
@@ -0,0 +1,2 @@
+$('#follow_form').html('<%= escape_javascript(render('users/unfollow')) %>');
```
Comment (chungpv-1008) [inline app/views/relationships/create.js.erb]:
```suggestion
$('#follow_form').html('<%= j render 'users/unfollow' %>');
```

File: app/views/relationships/destroy.js.erb [hunk 17]
```diff
@@ -0,0 +1,2 @@
+$('#follow_form').html('<%= escape_javascript(render('users/follow')) %>');
```
Comment (chungpv-1008) [inline app/views/relationships/destroy.js.erb]:
```suggestion
$('#follow_form').html('<%= j render 'users/follow' %>');
```

File: db/schema.rb [hunk 18]
```diff
@@ -10,7 +10,8 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2020_09_09_090813) do
+ActiveRecord::Schema.define(version: 2020_09_10_073120) do
+>>>>>>> chapter 14
```
Comment (chungpv-1008) [inline db/schema.rb]:
```suggestion

```

**Comment by baaohuii** [comment]
ready

**Review by chungpv-1008 — COMMENTED** [review]
merged

**Pull request merged**